### ISSUE TITLE
Nested documents add an extra line break?

### ISSUE THREAD

**Issue by edmcman** [afranchuk/prettyprint#3]
[This code](https://github.com/edmcman/prettyprint/blob/bug2/example/c-api.c#L64) prints:

```
[[[
  
  stuff st
  uff stuf
  f
]]]
```

I am not sure where the empty line is coming from.  If I understand correctly, nesting should not introduce a new line break.  So there should only be one line break coming from the pp_line call on https://github.com/edmcman/prettyprint/blob/bug2/example/c-api.c#L57.

**Comment by edmcman** [comment]
The extra newline is coming from the pp_text: https://github.com/edmcman/prettyprint/blob/bug2/src/prettyprint_base.c#L99

This seems wrong to me as a user, because if I wanted a line break to occur there, I would have added a `pp_line()`.  Since I did add a `pp_line()`, I'm now getting two line breaks.

**Comment by afranchuk** [comment]
Yeah you may be right. I believe the motivation for that is to try to wrap an entire word rather than immediately truncating it, giving the best chance of fitting more text of the word on a line without truncation. Maybe something could be done to check whether inserting a new line would give any more space (which would fix the case you hit).

**Comment by afranchuk** [comment]
In particular this sort of algorithm will avoid truncation for any word that's smaller than the print width, which is desirable for readability. Without that, you'd get truncation all the time and the text would be fairly unreadable.

**Comment by afranchuk** [comment]
Note this is also what `pp_words` is for; in the new test you added you use `pp_string("stuff stuff stuff")` which is why it splits in the middle of the words. But now that I think about it, `pp_words` could now be implemented to add a break at the start with no space as the replacement text (using the new feature you've added), which I think would have the same behavior. If so, that might be the cleanest approach.

**Comment by edmcman** [comment]
I find it a little confusing that `pp_words` and `pp_string` can both break strings up in different ways.  Also, I don't think the splitting behavior of `pp_string` is described anywhere.  FWIW, after reading Wadler's paper, I would expect that strings are never split, and a helper function like `pp_words` would do something like `group (line* + word1 + line + word2 + line)`, which I think is what you are suggesting.

Are you also suggesting to change `pp_string` to not split?

**Comment by afranchuk** [comment]
Well I wouldn't equate those two: `pp_words` is just internally using `pp_string` and `pp_line`, whereas `pp_string` is an atomic document type.

Yes, I'm suggesting to change `pp_words` to add the leading line and change `pp_string` to not have that behavior. I think I added that behavior for better readability, probably only because of the lack of a `line*`.
